bluemira/radiation_transport/radiation_profile.py

```diff
@@ -34,13 +34,627 @@
+        upstream temperature. Unit [keV]
+    q_u: float
+        upstream power density [W/m^2]
+    """
+    # minor radius
+    a = R_0 / a_ratio
+    p_rad = constants.raw_uc(p_rad, "MW", "W")
+
+    # SoL cross-section at the midplane
+    a_par = 4 * np.pi * a * (k ** (1 / 2)) * n * lambda_q_f
+    # power density at the upstream
+    q_u = (p_rad * q_95) / a_par
+
+    # connection length from the midplane to the target
+    l_tot = calculate_connection_length_flt(
+        eq,
+        r_sep_omp,
+        z_mp,
+        first_wall=firstwall_geom,
+    )
+
+    # upstream temperature [keV]
+    t_upstream_ev = (3.5 * (q_u / k_0) * l_tot) ** (2 / 7)
+    t_upstream_kev = constants.raw_uc(t_upstream_ev, "eV", "keV")
+
+    return t_upstream_kev, q_u
+
+
+def target_temperature(
+    q_u, t_u, lfs_p_fraction, div_p_sharing, n_el_0, gamma, eps_cool, f_ion_t, lfs=True
+):
+    """
+    Calculation of the temperature at the target location.
+    Extended 2PM. It includes hydrogen recycle loss energy.
+    Ref. Stangeby, "The Plasma Boundary of Magnetic Fusion
+    Devices", 2000.
+
+    Parameters
+    ----------
+    q_u: float
+        Upstream power density [W/m^2]
+    t_u: float
+        Upstream temperature. Unit [keV]
+    lfs_p_fraction: float
+        lfs fraction of rad power
+    div_p_sharing: float
+        Power fraction towards each divertor
+    n_el_0: float
+        Electron density on axis [1/m^3]
+    gamma: float
+        Sheath heat transmission coefficient
+    eps_cool: float
+        Electron energy loss [eV]
+    f_ion_t: float
+        Hydrogen first ionization [eV]
+    lfs: boolean
+        low field side. Default value True.
+        If False it stands for high field side (hfs).
+
+    Returns
+    -------
+    t_tar: float
+        target temperature. Unit [keV]
+    """
+    p_fraction = lfs_p_fraction if lfs else 1 - lfs_p_fraction
+    q_u = p_fraction * q_u * div_p_sharing
+    # Speed of light to convert kg to eV/c^2
+    light_speed = constants.C_LIGHT
+    # deuterium ion mass
+    m_i_amu = constants.D_MOLAR_MASS
+    m_i_kg = constants.raw_uc(m_i_amu, "amu", "kg")
+    m_i = m_i_kg / (light_speed**2)
+    # From keV to eV
+    t_u = constants.raw_uc(t_u, "keV", "eV")
+    n_u = n_el_0
+    # Numerator and denominator of the upstream forcing function
+    num_f = m_i * 4 * (q_u**2)
+    den_f = (
+        2
+        * constants.E_CHARGE
+        * (gamma**2)
+        * (constants.E_CHARGE**2)
+        * (n_u**2)
+        * (t_u**2)
+    )
+    # To address all the conversion from J to eV
+    f_ev = constants.raw_uc(num_f / den_f, "J", "eV")
+    # Critical target temperature
+    t_crit = eps_cool / gamma
+    # Finding roots of the target temperature quadratic equation
+    coeff_2 = 2 * (eps_cool / gamma) - f_ev
+    coeff_3 = (eps_cool**2) / (gamma**2)
+    coeff = [1, coeff_2, coeff_3]
+    roots = np.roots(coeff)
+    if roots.dtype == complex:
+        t_tar = constants.raw_uc(f_ion_t, "keV", "eV")
```

Comment:
this logic here could be joined with line 193 making this just `t_tar = f_ion_t` and tabbing in l193